graphics/src/widget/rule.rs

```diff
@@ -0,0 +1,152 @@
+/// This is an alias of an `iced_native` rule with an `iced_graphics::Renderer`.
+pub type Rule<Backend> = iced_native::Rule<Renderer<Backend>>;
+
+impl<B> rule::Renderer for Renderer<B>
+where
+    B: Backend,
+{
+    type Style = Box<dyn StyleSheet>;
+
+    fn draw(
+        &mut self,
+        bounds: Rectangle,
+        style_sheet: &Self::Style,
+        is_horizontal: bool,
+    ) -> Self::Output {
+        let style = style_sheet.style();
+
+        let line = if is_horizontal {
+            let line_y = (bounds.y + (bounds.height / 2.0)
+                - (style.width as f32 / 2.0))
+                .round();
+
+            let (line_x, line_width) = match style.fill_mode {
+                FillMode::Full => (bounds.x, bounds.width),
+                FillMode::Percent(percent) => {
+                    if percent >= 100.0 {
+                        (bounds.x, bounds.width)
+                    } else {
+                        let percent_width =
+                            (bounds.width * percent / 100.0).round();
+
+                        (
+                            bounds.x
+                                + ((bounds.width - percent_width) / 2.0)
+                                    .round(),
+                            percent_width,
+                        )
+                    }
+                }
+                FillMode::Padded(padding) => {
+                    if padding == 0 {
+                        (bounds.x, bounds.width)
+                    } else {
+                        let padding = padding as f32;
+                        let mut line_width = bounds.width - (padding * 2.0);
+                        if line_width < 0.0 {
+                            line_width = 0.0;
+                        }
+
+                        (bounds.x + padding, line_width)
+                    }
+                }
+                FillMode::AsymmetricPadding(first_pad, second_pad) => {
+                    let first_pad = first_pad as f32;
+                    let second_pad = second_pad as f32;
+                    let mut line_width = bounds.width - first_pad - second_pad;
+                    if line_width < 0.0 {
+                        line_width = 0.0;
+                    }
+
+                    (bounds.x + first_pad, line_width)
+                }
+            };
+
+            Primitive::Quad {
+                bounds: Rectangle {
+                    x: line_x,
+                    y: line_y,
+                    width: line_width,
+                    height: style.width as f32,
+                },
+                background: Background::Color(style.color),
+                border_radius: style.radius,
+                border_width: 0,
+                border_color: Color::TRANSPARENT,
+            }
+        } else {
+            let line_x = (bounds.x + (bounds.width / 2.0)
+                - (style.width as f32 / 2.0))
+                .round();
+
+            let (line_y, line_height) = match style.fill_mode {
```

Comment:
It looks like this logic can be abstracted away. Maybe a `FillMode::fill(space: f32) -> (f32, f32)` method returning the starting `offset` and the `length` of the `Rule`.

We should be able to use it for both horizontal and vertical rules.

Reply:
Good catch! Sometimes I forget you can add member methods to enums. I could probably clean up a lot of code in `iced_audio` with them as well :).